presets/presetManager.py

```diff
@@ -0,0 +1,272 @@
+import os.path
+import logging
+import json
+
+logger = logging.getLogger(__name__)
+
+
+class PresetManager(object):
+    '''PresetManager deals with presets loading, validating, storing
+
+    you can use it like this:
+
+    python::
+    pm = PresetManager(["/path/to/presets/folder", "/another/path"])
+    '''
+
+    MAX_DEPTH = 2
+
+    def __init__(self, paths, strict=False):
+        """load, validate, and store presets under ```self.presets```
+
+        :param paths:  list of paths to preset file
+                       or folder containing presets.
+        :param strict: if strict is False
+                       all exceptions and errors
+                       will be converted to log messages.
+        """
+
+        self.presets = {}
+        self.strict = strict
+
+        # normalize paramter type string -> array
+        if isinstance(paths, basestring):
+            paths = [paths]
+
+        self._loadPaths(paths)
+
+    def _loadPaths(self, paths, depth=0):
+        '''
+        Goes recursevly through the given list of paths
+        in order to find and pass all preset files to ```_loadPreset()```
+        '''
+
+        if depth > self.MAX_DEPTH:
+            return
+
+        for p in paths:
+            path = p
```

Comment:
>For strings seems to be different:

Strings are immutable, you can't expect them to change in any case.

>    path = os.path.normpath(path)

Indeed, you are not even trying to change them, you are just making the
variable point to something else

I think that you should better understand how variables works in Python.
They're actually more similar to names than to "what C family calls variables"

When you do
```
a = "foo"
b = a
a = "foozah"
```
what are you expecting to see?

>Both for cycles do not modify objects in ```l```.

...But we're losing the focus. I was wondering why you wrote such code.
I proposed the hypothesis that you want to avoid side-effects, and
explained you that even in that case, the technique you used was giving
no guarantees in that respect. Now you're pointing out yet another
reason why there are no side effects. Ok. But then why is the code
written like that?  Every time I read it, I feel dumb :(

A more concrete proposal:
```python
normpaths = (os.path.expanduser(os.path.expandvars(os.path.normpath(p))) for p in paths if p)

for path in normpaths:
    #all the rest of the code
```

I think this is shorter, clearly has no side-effect, separate the normalization part from the loading part. There are probably many other ways to write the same thing, find the one that you like more.